Golangci-lint does not show revive `exported` lints when using `import "C"`

### Welcome

- [x] Yes, I'm using a binary release within 2 latest major releases. Only such installations are supported.
- [X] Yes, I've searched similar issues on GitHub and didn't find any.
- [X] Yes, I've included all information below (version, config, etc).
- [X] Yes, I've tried with the standalone linter if available. (https://golangci-lint.run/usage/linters/)

### Description of the problem

When I enable the `revive` linter along with setting the `include` configuration to include exported lints the linter output fails when `import "C"` is used.

I've been using cgo to call into SDL2 and I noticed today that in files where I `import "C"` and `#include` a `.h` file, the exported class of lints (doc comments on exported function etc.) don't show up.

These lints work prefectly fine if I don't `import "C"`.

The lints also work fine if I enable the now deprecated `golint` linter even without removing `import "C"`, so this is just a problem with `revive`, for some reason.

### Version of golangci-lint

<details>

```console
$ golangci-lint --version
golangci-lint has version 1.43.0 built from 861262b7 on 2021-11-03T11:57:46Z
```

</details>

### Configuration file

<details>

```console
$ cat .golangci.json
{
    "linters": {
        "disable-all": true,
        "enable": [
            "revive"
        ]
    },
    "issues": {
        "include": [
            "EXC0012",
            "EXC0013",
            "EXC0014",
            "EXC0015"
        ]
    }
}
```

</details>

### Go environment

<details>

```console
$ go version && go env
go version go1.17.1 windows/amd64
set GO111MODULE=
set GOARCH=amd64
set GOBIN=D:\projects\go\bin
set GOCACHE=D:\devenv\go\1.17.1\..\cache
set GOENV=C:\Users\Polyscone\AppData\Roaming\go\env
set GOEXE=.exe
set GOEXPERIMENT=
set GOFLAGS=
set GOHOSTARCH=amd64
set GOHOSTOS=windows
set GOINSECURE=
set GOMODCACHE=D:\projects\go\pkg\mod
set GONOPROXY=
set GONOSUMDB=
set GOOS=windows
set GOPATH=D:\projects\go
set GOPRIVATE=
set GOPROXY=https://proxy.golang.org,direct
set GOROOT=D:\devenv\go\1.17.1
set GOSUMDB=sum.golang.org
set GOTMPDIR=
set GOTOOLDIR=D:\devenv\go\1.17.1\pkg\tool\windows_amd64
set GOVCS=
set GOVERSION=go1.17.1
set GCCGO=gccgo
set AR=ar
set CC=gcc
set CXX=g++
set CGO_ENABLED=1
set GOMOD=C:\Users\Polyscone\Desktop\bug\go.mod
set CGO_CFLAGS=-g -O2
set CGO_CPPFLAGS=
set CGO_CXXFLAGS=-g -O2
set CGO_FFLAGS=-g -O2
set CGO_LDFLAGS=-g -O2
set PKG_CONFIG=pkg-config
set GOGCCFLAGS=-m64 -mthreads -fmessage-length=0 -fdebug-prefix-map=C:\Users\POLYSC~1\AppData\Local\Temp\go-build1279465700=/tmp/go-build -gno-record-gcc-switches
```

</details>

### Verbose output of running

Without `import "C"`:
<details>

```console
$ golangci-lint cache clean
$ golangci-lint run -v
level=info msg="[config_reader] Config search paths: [./ C:\\Users\\Polyscone\\Desktop\\bug C:\\Users\\Polyscone\\Desktop C:\\Users\\Polyscone C:\\Users C:\\ D:\\devenv]"
level=info msg="[config_reader] Used config file .golangci.json"
level=info msg="[lintersdb] Active 1 linters: [revive]"
level=info msg="[loader] Go packages loading at mode 7 (name|compiled_files|files) took 508.7361ms"
level=info msg="[runner/filename_unadjuster] Pre-built 0 adjustments in 0s"
level=info msg="[linters context/goanalysis] analyzers took 0s with top 10 stages: the_only_name: 0s"
level=info msg="[runner] Processors filtering stat (out/in): skip_dirs: 1/1, cgo: 1/1, path_prettifier: 1/1, exclude: 1/1, nolint: 1/1, max_per_file_from_linter: 1/1, max_same_issues: 1/1, path_prefixer: 1/1, filename_unadjuster: 1/1, diff: 1/1, max_from_linter: 1/1, path_shortener: 1/1, sort_results: 1/1, skip_files: 1/1, autogenerated_exclude: 1/1, identifier_marker: 1/1, exclude-rules: 1/1, uniq_by_line: 1/1, source_code: 1/1, severity-rules: 1/1"
level=info msg="[runner] processing took 0s with stages: autogenerated_exclude: 0s, identifier_marker: 0s, exclude: 0s, nolint: 0s, max_per_file_from_linter: 0s, cgo: 0s, path_prettifier: 0s, exclude-rules: 0s, max_from_linter: 0s, path_shortener: 0s, severity-rules: 0s, skip_files: 0s, skip_dirs: 0s, source_code: 0s, filename_unadjuster: 0s, uniq_by_line: 0s, diff: 0s, max_same_issues: 0s, path_prefixer: 0s, sort_results: 0s"
level=info msg="[runner] linters took 3.9982ms with stages: revive: 3.0007ms"
sdl\sdl.go:3:1: exported: exported function Foo should have comment or be unexported (revive)
func Foo() {}
^
level=info msg="File cache stats: 1 entries of total size 27B"
level=info msg="Memory: 8 samples, avg is 44.1MB, max is 44.4MB"
level=info msg="Execution took 629.5749ms"
```

</details>

With `import "C"`:
<details>

```console
$ golangci-lint cache clean
$ golangci-lint run -v
level=info msg="[config_reader] Config search paths: [./ C:\\Users\\Polyscone\\Desktop\\bug C:\\Users\\Polyscone\\Desktop C:\\Users\\Polyscone C:\\Users C:\\ D:\\devenv]"
level=info msg="[config_reader] Used config file .golangci.json"
level=info msg="[lintersdb] Active 1 linters: [revive]"
level=info msg="[loader] Go packages loading at mode 7 (files|name|compiled_files) took 570.3007ms"
level=info msg="[runner/filename_unadjuster] Pre-built 0 adjustments in 971.9µs"
level=info msg="[linters context/goanalysis] analyzers took 2.0046ms with top 10 stages: the_only_name: 2.0046ms"
level=info msg="[runner] processing took 0s with stages: uniq_by_line: 0s, sort_results: 0s, autogenerated_exclude: 0s, skip_dirs: 0s, exclude-rules: 0s, diff: 0s, max_per_file_from_linter: 0s, max_same_issues: 0s, path_shortener: 0s, severity-rules: 0s, cgo: 0s, path_prettifier: 0s, identifier_marker: 0s, exclude: 0s, path_prefixer: 0s, filename_unadjuster: 0s, nolint: 0s, max_from_linter: 0s, source_code: 0s, skip_files: 0s"
level=info msg="[runner] linters took 551.2396ms with stages: revive: 551.2396ms"
level=info msg="File cache stats: 0 entries of total size 0B"
level=info msg="Memory: 14 samples, avg is 48.0MB, max is 48.0MB"
level=info msg="Execution took 1.2283984s"
```

</details>

### Code example or link to a public repository

The minimum project I set up is a `.golangci.json` file with the above config, a `go.mod` file, a `main.go` file with an empty `main` function in it, and then an `sdl` package.

The `sdl` package contains a single `sdl.go` file with the following contents:

<details>

```go
package sdl

// #include <SDL.h>
import "C"

func Foo() {}
```

</details>

As soon as I remove the `import "C"` golangci-lint correctly reports the missing doc comments.

--- a/pkg/golinters/dupl.go
+++ b/pkg/golinters/dupl.go
@@ -55,11 +55,7 @@ func NewDupl(settings *config.DuplSettings) *goanalysis.Linter {
 }
 
 func runDupl(pass *analysis.Pass, settings *config.DuplSettings) ([]goanalysis.Issue, error) {
-	var fileNames []string
-	for _, f := range pass.Files {
-		pos := pass.Fset.PositionFor(f.Pos(), false)
-		fileNames = append(fileNames, pos.Filename)
-	}
+	fileNames := getFileNames(pass)
 
 	issues, err := duplAPI.Run(fileNames, settings.Threshold)
 	if err != nil {

--- a/pkg/golinters/gci.go
+++ b/pkg/golinters/gci.go
@@ -83,11 +83,7 @@ func NewGci(settings *config.GciSettings) *goanalysis.Linter {
 }
 
 func runGci(pass *analysis.Pass, lintCtx *linter.Context, cfg *gcicfg.Config, lock *sync.Mutex) ([]goanalysis.Issue, error) {
-	var fileNames []string
-	for _, f := range pass.Files {
-		pos := pass.Fset.PositionFor(f.Pos(), false)
-		fileNames = append(fileNames, pos.Filename)
-	}
+	fileNames := getFileNames(pass)
 
 	var diffs []string
 	err := diffFormattedFilesToArray(fileNames, *cfg, &diffs, lock)

--- a/pkg/golinters/gofmt.go
+++ b/pkg/golinters/gofmt.go
@@ -53,11 +53,7 @@ func NewGofmt(settings *config.GoFmtSettings) *goanalysis.Linter {
 }
 
 func runGofmt(lintCtx *linter.Context, pass *analysis.Pass, settings *config.GoFmtSettings) ([]goanalysis.Issue, error) {
-	var fileNames []string
-	for _, f := range pass.Files {
-		pos := pass.Fset.PositionFor(f.Pos(), false)
-		fileNames = append(fileNames, pos.Filename)
-	}
+	fileNames := getFileNames(pass)
 
 	var issues []goanalysis.Issue
 

--- a/pkg/golinters/gofumpt.go
+++ b/pkg/golinters/gofumpt.go
@@ -73,11 +73,7 @@ func NewGofumpt(settings *config.GofumptSettings) *goanalysis.Linter {
 }
 
 func runGofumpt(lintCtx *linter.Context, pass *analysis.Pass, diff differ, options format.Options) ([]goanalysis.Issue, error) {
-	var fileNames []string
-	for _, f := range pass.Files {
-		pos := pass.Fset.PositionFor(f.Pos(), false)
-		fileNames = append(fileNames, pos.Filename)
-	}
+	fileNames := getFileNames(pass)
 
 	var issues []goanalysis.Issue
 

--- a/pkg/golinters/goimports.go
+++ b/pkg/golinters/goimports.go
@@ -55,11 +55,7 @@ func NewGoimports(settings *config.GoImportsSettings) *goanalysis.Linter {
 }
 
 func runGoiImports(lintCtx *linter.Context, pass *analysis.Pass) ([]goanalysis.Issue, error) {
-	var fileNames []string
-	for _, f := range pass.Files {
-		pos := pass.Fset.PositionFor(f.Pos(), false)
-		fileNames = append(fileNames, pos.Filename)
-	}
+	fileNames := getFileNames(pass)
 
 	var issues []goanalysis.Issue
 

--- a/pkg/golinters/gomodguard.go
+++ b/pkg/golinters/gomodguard.go
@@ -73,12 +73,7 @@ func NewGomodguard(settings *config.GoModGuardSettings) *goanalysis.Linter {
 		}
 
 		analyzer.Run = func(pass *analysis.Pass) (interface{}, error) {
-			var files []string
-			for _, file := range pass.Files {
-				files = append(files, pass.Fset.PositionFor(file.Pos(), false).Filename)
-			}
-
-			gomodguardIssues := processor.ProcessFiles(files)
+			gomodguardIssues := processor.ProcessFiles(getFileNames(pass))
 
 			mu.Lock()
 			defer mu.Unlock()

--- a/pkg/golinters/lll.go
+++ b/pkg/golinters/lll.go
@@ -56,11 +56,7 @@ func NewLLL(settings *config.LllSettings) *goanalysis.Linter {
 }
 
 func runLll(pass *analysis.Pass, settings *config.LllSettings) ([]goanalysis.Issue, error) {
-	var fileNames []string
-	for _, f := range pass.Files {
-		pos := pass.Fset.PositionFor(f.Pos(), false)
-		fileNames = append(fileNames, pos.Filename)
-	}
+	fileNames := getFileNames(pass)
 
 	spaces := strings.Repeat(" ", settings.TabWidth)
 

--- a/pkg/golinters/misspell.go
+++ b/pkg/golinters/misspell.go
@@ -61,12 +61,7 @@ func NewMisspell(settings *config.MisspellSettings) *goanalysis.Linter {
 }
 
 func runMisspell(lintCtx *linter.Context, pass *analysis.Pass, replacer *misspell.Replacer) ([]goanalysis.Issue, error) {
-	var fileNames []string
-
-	for _, f := range pass.Files {
-		pos := pass.Fset.PositionFor(f.Pos(), false)
-		fileNames = append(fileNames, pos.Filename)
-	}
+	fileNames := getFileNames(pass)
 
 	var issues []goanalysis.Issue
 	for _, filename := range fileNames {

--- a/pkg/golinters/revive.go
+++ b/pkg/golinters/revive.go
@@ -75,11 +75,7 @@ func NewRevive(settings *config.ReviveSettings) *goanalysis.Linter {
 }
 
 func runRevive(lintCtx *linter.Context, pass *analysis.Pass, settings *config.ReviveSettings) ([]goanalysis.Issue, error) {
-	var files []string
-	for _, file := range pass.Files {
-		files = append(files, pass.Fset.PositionFor(file.Pos(), false).Filename)
-	}
-	packages := [][]string{files}
+	packages := [][]string{getFileNames(pass)}
 
 	conf, err := getReviveConfig(settings)
 	if err != nil {

--- a/pkg/golinters/util.go
+++ b/pkg/golinters/util.go
@@ -2,7 +2,10 @@ package golinters
 
 import (
 	"fmt"
+	"path/filepath"
 	"strings"
+
+	"golang.org/x/tools/go/analysis"
 
 	"github.com/golangci/golangci-lint/pkg/config"
 )
@@ -21,4 +24,18 @@ func formatCodeBlock(code string, _ *config.Config) string {
 	}
 
 	return fmt.Sprintf("```\n%s\n```", code)
+}
+
+func getFileNames(pass *analysis.Pass) []string {
+	var fileNames []string
+	for _, f := range pass.Files {
+		fileName := pass.Fset.PositionFor(f.Pos(), true).Filename
+		ext := filepath.Ext(fileName)
+		if ext != "" && ext != ".go" {
+			// position has been adjusted to a non-go file, revert to original file
+			fileName = pass.Fset.PositionFor(f.Pos(), false).Filename
+		}
+		fileNames = append(fileNames, fileName)
+	}
+	return fileNames
 }

--- a/pkg/golinters/wsl.go
+++ b/pkg/golinters/wsl.go
@@ -67,15 +67,11 @@ func NewWSL(settings *config.WSLSettings) *goanalysis.Linter {
 }
 
 func runWSL(pass *analysis.Pass, conf *wsl.Configuration) []goanalysis.Issue {
-	var files = make([]string, 0, len(pass.Files))
-	for _, file := range pass.Files {
-		files = append(files, pass.Fset.PositionFor(file.Pos(), false).Filename)
-	}
-
 	if conf == nil {
 		return nil
 	}
 
+	files := getFileNames(pass)
 	wslErrors, _ := wsl.NewProcessorWithConfig(*conf).ProcessFiles(files)
 	if len(wslErrors) == 0 {
 		return nil

--- a/test/run_test.go
+++ b/test/run_test.go
@@ -104,6 +104,8 @@ func TestCgoWithIssues(t *testing.T) {
 		ExpectHasIssue("Printf format %t has arg cs of wrong type")
 	r.Run("--no-config", "--disable-all", "-Estaticcheck", getTestDataDir("cgo_with_issues")).
 		ExpectHasIssue("SA5009: Printf format %t has arg #1 of wrong type")
+	r.Run("--no-config", "--disable-all", "-Egofmt", getTestDataDir("cgo_with_issues")).
+		ExpectHasIssue("File is not `gofmt`-ed with `-s` (gofmt)")
 }
 
 func TestUnsafeOk(t *testing.T) {

--- a/test/run_test.go
+++ b/test/run_test.go
@@ -106,6 +106,8 @@ func TestCgoWithIssues(t *testing.T) {
 		ExpectHasIssue("SA5009: Printf format %t has arg #1 of wrong type")
 	r.Run("--no-config", "--disable-all", "-Egofmt", getTestDataDir("cgo_with_issues")).
 		ExpectHasIssue("File is not `gofmt`-ed with `-s` (gofmt)")
+	r.Run("--no-config", "--disable-all", "-Erevive", getTestDataDir("cgo_with_issues")).
+		ExpectHasIssue("indent-error-flow: if block ends with a return statement")
 }
 
 func TestUnsafeOk(t *testing.T) {
@@ -129,7 +131,7 @@ func TestLineDirectiveProcessedFilesLiteLoading(t *testing.T) {
 }
 
 func TestSortedResults(t *testing.T) {
-	var testCases = []struct {
+	testCases := []struct {
 		opt  string
 		want string
 	}{

--- a/test/run_test.go
+++ b/test/run_test.go
@@ -95,7 +95,7 @@ func TestTestsAreLintedByDefault(t *testing.T) {
 }
 
 func TestCgoOk(t *testing.T) {
-	testshared.NewLintRunner(t).Run("--no-config", "--enable-all", "-D", "nosnakecase", getTestDataDir("cgo")).ExpectNoIssues()
+	testshared.NewLintRunner(t).Run("--no-config", "--enable-all", "-D", "nosnakecase,gci", getTestDataDir("cgo")).ExpectNoIssues()
 }
 
 func TestCgoWithIssues(t *testing.T) {

--- a/test/testdata/cgo_with_issues/main.go
+++ b/test/testdata/cgo_with_issues/main.go
@@ -21,3 +21,15 @@ func Example() {
 	fmt.Printf("bad format %t", cs)
 	C.free(unsafe.Pointer(cs))
 }
+
+func notFormattedForGofmt()  {
+}
+
+func errorForRevive(p *int) error {
+	if p == nil {
+		return nil
+	} else {
+		return nil
+	}
+}
+

